test/gtest/ucp/test_ucp_wireup.cc

```diff
@@ -1149,6 +1205,60 @@ class test_ucp_wireup_fallback_amo : public test_ucp_wireup {
     }
 };
 
+class test_ucp_wireup_amo : public test_ucp_wireup {
+public:
+    static ucp_params_t get_ctx_params() {
+        ucp_params_t params = test_ucp_wireup::get_ctx_params();
+        params.field_mask  |= UCP_PARAM_FIELD_REQUEST_SIZE;
+        params.request_size = sizeof(test_ucp_wireup_amo *);
+        return params;
+    }
+
+    static std::vector<ucp_test_param>
+    enum_test_params(const ucp_params_t& ctx_params, const std::string& name,
+                     const std::string& test_case_name, const std::string& tls)
+    {
+        return enum_test_params_features(ctx_params, name, test_case_name, tls,
+                                         UCP_FEATURE_AMO32 | UCP_FEATURE_AMO64,
+                                         false);
+    }
+
+protected:
+    static void flush_cb(void *req, ucs_status_t status) {
+        test_ucp_wireup_amo **test_p = (test_ucp_wireup_amo **)
+                       UCS_PTR_BYTE_OFFSET(req, -sizeof(test_ucp_wireup_amo *));
+        test_ucp_wireup_amo *self    = *test_p;
+        self->rkeys_cleanup();
+        self->memhs_cleanup();
+    }
+};
+
+UCS_TEST_P(test_ucp_wireup_amo, relese_key_after_flush) {
+    fill_send_data();
+    clear_recv_data();
+
+    sender().connect(&receiver(), get_ep_params());
+
+    ucp_rkey_h rkey = get_rkey(sender());
+    add_rkey(rkey);
+
+    ucs_status_t status = ucp_atomic_post(sender().ep(), UCP_ATOMIC_POST_OP_ADD,
+                                          get_amo_value(), sizeof(elem_type),
+                                          get_amo_recv_addr(), rkey);
+    ASSERT_UCS_OK(status);
+    void *req = ucp_ep_flush_nb(sender().ep(), UCT_FLUSH_FLAG_LOCAL, flush_cb);
+    if (UCS_PTR_IS_PTR(req)) {
+        test_ucp_wireup_amo **req_headroom = (test_ucp_wireup_amo **)
```

Comment:
seems wrong, the user data is after UCP, not before, this seems to actually override ucp_request_t portion